docs/design/data_plane_manager.adoc

```diff
@@ -0,0 +1,1189 @@
+It provides a flexible plugin framework which allows different data-plane extension.
+
+=== User Cases
+
+{empty}(1) Port Manger wants to send one new port information to every hosts in the VPC +
+(2) Security Group Manager wants to send one new security group to hosts in the VPC
+
+== Service Requirements
+
+=== Basic Requirements
+
+[arabic]
+. Provide a flexible plugin framework which allows different data-plane extensions.
+. OVS extension supports integration with OVS
+. Reliably connecting to hosts through various paths including normal path (kafka) and fast path (gPRC)
+. Configurable retry mechanism when connection to hosts is unstable
+. Working together with other services including port manager and security group manager.
+. Concurrency control mechanism should be provided to process multiple concurrent network configuration programming requests.
+
+=== Advanced Requirements
+
+Fall back mechanism from one path to the other, for example, from fast path to normal path, or from normal path to rescue path
+
+== Data-Plane Microservice Design
+
+=== Architecture
+
+image:images/dpm.png[image,width=660,height=384]
+
+[Figure 1] Data Plane Manager Structure
+
+=== Extensible Design to Support Various Data-Plane
+
+Upstream microservices doesn't need to know about the actual data-plane implementation.
+
+. OVS Extension
+. Mizar Gateway Extension (later)
+
+=== Key Worfklow
+
+. Create first port in a subnet
+
+image:images/dpm2.png[image,width=660,height=384]
+
+[Figure 2] create first port 
+
+image:images/dpm3.png[image,width=660,height=384]
+
+[Figure 3] create first port 
+
+. Create a port in an existing subnet
+. Update a port
+. Get a port
+. Delete a port
+
+=== Concurrency Control
+
+Data plane manger service would do the provision work concurrently using multithread technology.
+
+== Communication with Agent
+
+=== Message Design
+* For initial data we would do full update with nfs* support and notification on messaging Q
+* We would use delta update for small amount of update to deploy
+* NFS could support up to 30G/s speed and we could use advanced nfs to get half of response time with bypassing kernel 
+
+=== Multi-path Control Logic
+
+. Normal Path
+. Fast Path
+. Rescue Path
+
+Logic:
+
+. Initially we would use fast path (grpc) to tell agent the info about next step
+. We would normally use normal path to do agent provision and nfs for big file (3GB file with gzip would cost about 3s)
+. When heart beat loss decteted (via Netdata) , data plane manager would send out rescue message to MessagingQ to let adjacent ovs do remote rescue
+
+==== Problems solution:
+
+. To solve the misorder data, put them in same queue
+. Use the tag to filter msg.
+. RocketMQ could support way more topic number with stable performance, we need to fine tune the parameter for dirty page and memory usage
+. Use bond to get more bandwidth for storm
```

Comment:
Give a bit more information on what bond could do to help this case.

Reply:
Network bonding is a process of combing or joining two or more network interfaces together into a single interface. Network bonding offers performance improvements and redundancy by increasing the network throughput and bandwidth. If one interface is down or unplugged the other one will work. It can be used in situations where you need fault tolerance, redundancy or load balancing networks. In Linux, a special kernel module named bonding is used to connect multiple network interfaces into a single interface. Two or more network interfaces can be connected into a single logical “bonded” interface. The behavior of the bonded interfaces depends on the type of bonding method used.

I would add it in the doc

Reply:
be careful of reliance of network bonding, I am worried that it may not work well with hardware offloading technologies like dpdk/sriov/smart nic etc.